@here/olp-sdk-dataservice-read/lib/utils/getTile.ts

```diff
@@ -61,109 +118,202 @@ import {
-    ).withQuadKey(quadKey);
-
-    if (layerType === "versioned") {
-        const catalogVersion = await rq
-            .getCatalogVersion()
-            .catch(err => Promise.reject(err));
-        quadTreeIndexRequest.withVersion(catalogVersion);
-    }
-
-    const quadTreeIndex = await queryClient
-        .fetchQuadTreeIndex(quadTreeIndexRequest, abortSignal)
-        .catch(err => Promise.reject(err));
+    const quadTreeIndex = await fetchQuadTreeIndex({
+        ...params,
+        catalogVersion,
+        depth: delta,
+        fetchOptions: request.getFetchOption(),
+        tileKey: parentTileKey,
+        abortSignal,
+        billingTag: request.getBillingTag()
+    }).catch(e => Promise.reject(e));
 
-    if (!quadTreeIndex || !quadTreeIndex.subQuads) {
+    if (
+        !quadTreeIndex ||
+        !quadTreeIndex.subQuads ||
+        !quadTreeIndex.subQuads.length
+    ) {
         return Promise.reject(new Error("Error fetching QuadTreeIndex"));
     }
 
-    if (
-        quadTreeIndex &&
-        quadTreeIndex.subQuads &&
-        quadTreeIndex.subQuads.length
-    ) {
-        if (fetchOptions !== FetchOptions.OnlineOnly) {
-            cache.put(partitionsRequest, catalogHrn.toString(), layerId, [
-                {
-                    partition: partitionId,
-                    dataHandle: quadTreeIndex.subQuads[0].dataHandle,
-                    version: quadTreeIndex.subQuads[0].version
-                }
-            ]);
-
-            if (quadTreeIndex.parentQuads && quadTreeIndex.parentQuads.length) {
-                cache.put(
-                    partitionsRequest,
-                    catalogHrn.toString(),
-                    layerId,
-                    quadTreeIndex.parentQuads
-                );
-            }
+    // Return the data for the requested QuadKey or for the closest parent
+    const subQuads = quadTreeIndex.subQuads;
+
+    let currentTileKey = requestedTileKey;
+    for (let level = currentTileKey.level; level >= 0; --level) {
+        const metadata = subQuads.find(item => item.subQuadKey === subQuadKey);
+
+        if (metadata) {
+            return BlobApi.getBlob(blobRequestBuilder, {
+                dataHandle: metadata.dataHandle,
+                layerId: params.layerId,
+                billingTag: request.getBillingTag()
+            });
         }
 
-        return BlobApi.getBlob(blobRequestBuilder, {
-            dataHandle: quadTreeIndex.subQuads[0].dataHandle,
-            layerId,
-            billingTag
-        });
+        try {
+            currentTileKey = currentTileKey.parent();
+        } catch (error) {
+            continue;
+        }
     }
 
     return Promise.reject(
         new Error(`Error getting blob for Tile: ${JSON.stringify(quadKey)}`)
     );
 }
+
+/**
+ * The function should request quad tree index for the parent tile with delta 4
+ * and cache the responses in cache for later calls.
+ * @hidden
+ */
+export async function fetchQuadTreeIndex(
```

Comment:
Is this method putting the quadtree index exploded into the cache, which means each partition for itself? Or does it store the entire quad tree as such?

Reply:
each partition for itself. We don't have implemented caching for entire quad tree. Should we do that?

Reply:
Not necessary, as long as you look for that partition in the cache before you try to query the quadtree index.

Reply:
yes, it works like that